README.md
xbmcaudiobiblebrowser
=====================

Browse bible recordings via XBMC


plugin.audio.audiobible/changelog.txt
0.0.1
- First Try
0.0.2
- First Working prototype
0.1.0
- All current features implemented


plugin.audio.audiobible/bibledata.py
#!/usr/bin/python
# -*- coding: utf-8 -*
import collections
recordinglib=dict({
#CH
'ccb-biblica': {'code': 'ccb-biblica', 'title': 'CCB by Biblica','section':'ALL','language':'CH','pluswisdom':0}, 
'csb-hao': {'code': 'csb-hao', 'title': 'CSB (New Testament)  by Ran Hao','section':'NT', 'language':'CH','pluswisdom':0}, 
#EN
'esv-laughlin': {'code': 'esv-laughlin', 'title': 'ESV (New Testament) by Marquis Laughlin', 'section':'NT', 'language':'EN','pluswisdom':0},
'esv-mclean': {'code': 'esv-mclean', 'title': 'ESV by Max McLean','section':'ALL','language':'EN','pluswisdom':0}, 
'hcsb-mcconachie': {'code': 'hcsb-mcconachie', 'title': 'HCSB by Dale McConachie','section':'ALL','language':'EN','pluswisdom':0}, 
'kjv-dramatized': {'code': 'kjv-dramatized', 'title': 'KJV by Dramatized','section':'ALL','language':'EN','pluswisdom':0}, 
'kjv-mclean': {'code': 'kjv-mclean', 'title': 'KJV by Max McLean','section':'ALL','language':'EN','pluswisdom':0}, 
'kjv-mims': {'code': 'kjv-mims', 'title': 'KJV (Old Testament)  by Paul Mims','section':'OT','language':'EN','pluswisdom':0}, 
'leb-logos': {'code': 'leb-logos', 'title': 'LEB (New Testament)  by Logos','section':'OT','language':'EN','pluswisdom':0}, 
'nasb-mcconachie': {'code': 'nasb-mcconachie', 'title': 'NASB by Dale McConachie','section':'ALL','language':'EN','pluswisdom':0}, 
'niv-dramatized': {'code': 'niv-dramatized', 'title': 'NIV by Dramatized','section':'ALL','language':'EN','pluswisdom':0}, 
'niv-mclean': {'code': 'niv-mclean', 'title': 'NIV by Max McLean','section':'ALL','language':'EN','pluswisdom':0},
'niv-purevoice': {'code': 'niv-purevoice', 'title': 'NIV by George W. Sarris','section':'ALL','language':'EN','pluswisdom':0}, 
#ES
'lbla-montoya': {'code': 'lbla-montoya', 'title': 'LBLA (New Testament)  by Samuel Montoya H','section':'NT','language':'ES','pluswisdom':0},
'nvi-single': {'code': 'nvi-single', 'title': 'NVI by Rafael Cruz','section':'ALL','language':'ES','pluswisdom':0}, 
#DE
'reimer-reimer': {'code': 'reimer-reimer', 'title': 'REIMER by Elmer Reimer','section':'NT','language':'DE','pluswisdom':1}, 
#SK
'tlm-hodul': {'code': 'tlm-hodul', 'title': 'SLO1979 (New Testament)  by Joseph Hodul','section':'NT','language':'SK','pluswisdom':0}, 
#CZ
'bk-hodul': {'code': 'bk-hodul', 'title': 'BK (New Testament)  by Joseph Hodul','section':'NT','language':'CZ','pluswisdom':0},
#FA
'tpv-as': {'code': 'tpv-as', 'title': 'Farsi by Audio Scriptures','section':'NT','language':'FA','pluswisdom':1}, 
#JP 
'jlb-biblica': {'code': 'jlb-biblica', 'title': 'JLB (New Testament)  by Biblica','section':'NT','language':'JP','pluswisdom':0},
#SW
'snt-biblica': {'code': 'snt-biblica', 'title': 'SNT (New Testament)  by Biblica','section':'NT','language':'SW','pluswisdom':0}
})
languagelib=dict({
'CH': {'title': '--- Chinese ---'}, 
'EN': {'title': '--- English ---'},
'ES': {'title': '--- Spanish ---'}, 
'DE': {'title': '--- German ---'}, 
'SK': {'title': '--- Slovak ---'}, 
'CZ': {'title': '--- Czech ---'}, 
'FA': {'title': '--- Farsi ---'}, 
'JP': {'title': '--- Japanese ---'},
'SW': {'title': '--- Swahili ---'}
})

pluswisdom=['reimer-reimer','tpv-as']

booklib=dict({'Gen': {'chapters':50, 'title':'Genesis', 'section':'OT'},
'Exod': {'chapters':40, 'title':'Exodus', 'section':'OT'},
'Lev': {'chapters':27, 'title':'Leviticus', 'section':'OT'},
'Num': {'chapters':36, 'title':'Numbers', 'section':'OT'},
'Deut': {'chapters':34, 'title':'Deuteronomy', 'section':'OT'},
'Josh': {'chapters':24, 'title':'Joshua', 'section':'OT'},
'Judg': {'chapters':21, 'title':'Judges', 'section':'OT'},
'Ruth': {'chapters':4, 'title':'Ruth', 'section':'OT'},
'1Sam': {'chapters':31, 'title':'1 Samuel', 'section':'OT'},
'2Sam': {'chapters':24, 'title':'2 Samuel', 'section':'OT'},
'1Kgs': {'chapters':22, 'title':'1 Kings', 'section':'OT'},
'2Kgs': {'chapters':25, 'title':'2 Kings', 'section':'OT'},
'1Chr': {'chapters':29, 'title':'1 Chronicles', 'section':'OT'},
'2Chr': {'chapters':36, 'title':'2 Chronicles', 'section':'OT'},
'Ezra': {'chapters':10, 'title':'Ezra', 'section':'OT'},
'Neh': {'chapters':13, 'title':'Nehemiah', 'section':'OT'},
'Esth': {'chapters':10, 'title':'Esther', 'section':'OT'},
'Job': {'chapters':42, 'title':'Job', 'section':'OT'},
'Ps': {'chapters':150, 'title':'Psalms', 'section':'OT'},
'Prov': {'chapters':31, 'title':'Proverbs', 'section':'OT'},
'Eccl': {'chapters':12, 'title':'Ecclesiastes', 'section':'OT'},
'Song': {'chapters':8, 'title':'Song of Solomon', 'section':'OT'},
'Isa': {'chapters':66, 'title':'Isaiah', 'section':'OT'},
'Jer': {'chapters':52, 'title':'Jeremiah', 'section':'OT'},
'Lam': {'chapters':5, 'title':'Lamentations', 'section':'OT'},
'Ezek': {'chapters':48, 'title':'Ezekiel', 'section':'OT'},
'Dan': {'chapters':12, 'title':'Daniel', 'section':'OT'},
'Hos': {'chapters':14, 'title':'Hosea', 'section':'OT'},
'Joel': {'chapters':3, 'title':'Joel', 'section':'OT'},
'Amos': {'chapters':9, 'title':'Amos', 'section':'OT'},
'Obad': {'chapters':1, 'title':'Obadiah', 'section':'OT'},
'Jonah': {'chapters':4, 'title':'Jonah', 'section':'OT'},
'Mic': {'chapters':7, 'title':'Micah', 'section':'OT'},
'Nah': {'chapters':3, 'title':'Nahum', 'section':'OT'},
'Hab': {'chapters':3, 'title':'Habakkuk', 'section':'OT'},
'Zeph': {'chapters':3, 'title':'Zephaniah', 'section':'OT'},
'Hag': {'chapters':2, 'title':'Haggai', 'section':'OT'},
'Zech': {'chapters':14, 'title':'Zechariah', 'section':'OT'},
'Mal': {'chapters':4, 'title':'Malachi', 'section':'OT'},
'Matt': {'chapters':28, 'title':'Matthew', 'section':'NT'},
'Mark': {'chapters':16, 'title':'Mark', 'section':'NT'},
'Luke': {'chapters':24, 'title':'Luke', 'section':'NT'},
'John': {'chapters':21, 'title':'John', 'section':'NT'},
'Acts': {'chapters':28, 'title':'Acts', 'section':'NT'},
'Rom': {'chapters':16, 'title':'Romans', 'section':'NT'},
'1Cor': {'chapters':16, 'title':'1 Corintians', 'section':'NT'},
'2Cor': {'chapters':13, 'title':'2 Corintians', 'section':'NT'},
'Gal': {'chapters':6, 'title':'Galatians', 'section':'NT'},
'Eph': {'chapters':6, 'title':'Ephesians', 'section':'NT'},
'Phil': {'chapters':4, 'title':'Philippians', 'section':'NT'},
'Col': {'chapters':4, 'title':'Colossians', 'section':'NT'},
'1Thess': {'chapters':5, 'title':'1 Thessalonians', 'section':'NT'},
'2Thess': {'chapters':3, 'title':'2 Thessalonians', 'section':'NT'},
'1Tim': {'chapters':6, 'title':'1 Timothy', 'section':'NT'},
'2Tim': {'chapters':4, 'title':'2 Timothy', 'section':'NT'},
'Titus': {'chapters':3, 'title':'Titus', 'section':'NT'},
'Phlm': {'chapters':1, 'title':'Philemon', 'section':'NT'},
'Heb': {'chapters':13, 'title':'Hebrews', 'section':'NT'},
'Jas': {'chapters':5, 'title':'James', 'section':'NT'},
'1Pet': {'chapters':5, 'title':'1 Peter', 'section':'NT'},
'2Pet': {'chapters':3, 'title':'2 Peter', 'section':'NT'},
'1John': {'chapters':5, 'title':'1 John', 'section':'NT'},
'2John': {'chapters':1, 'title':'2 John', 'section':'NT'},
'3John': {'chapters':1, 'title':'3 John', 'section':'NT'},
'Jude': {'chapters':1, 'title':'Jude', 'section':'NT'},
'Rev': {'chapters':22, 'title':'Revelation', 'section':'NT'}})

bookorderlib=dict({
1:'Gen',
2:'Exod',
3:'Lev',
4:'Num',
5:'Deut',
6:'Josh',
7:'Judg',
8:'Ruth',
9:'1Sam',
10:'2Sam',
11:'1Kgs',
12:'2Kgs',
13:'1Chr',
14:'2Chr',
15:'Ezra',
16:'Neh',
17:'Esth',
18:'Job',
19:'Ps',
20:'Prov',
21:'Eccl',
22:'Song',
23:'Isa',
24:'Jer',
25:'Lam',
26:'Ezek',
27:'Dan',
28:'Hos',
29:'Joel',
30:'Amos',
31:'Obad',
32:'Jonah',
33:'Mic',
34:'Nah',
35:'Hab',
36:'Zeph',
37:'Hag',
38:'Zech',
39:'Mal',
40:'Matt',
41:'Mark',
42:'Luke',
43:'John',
44:'Acts',
45:'Rom',
46:'1Cor',
47:'2Cor',
48:'Gal',
49:'Eph',
50:'Phil',
51:'Col',
52:'1Thess',
53:'2Thess',
54:'1Tim',
55:'2Tim',
56:'Titus',
57:'Phlm',
58:'Heb',
59:'Jas',
60:'1Pet',
61:'2Pet',
62:'1John',
63:'2John',
64:'3John',
65:'Jude',
66:'Rev'})

plugin.audio.audiobible/default.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
__addonname__     = "Audio Bible Browser"
__addonid__       = "plugin.audio.audiobible"
__author__        = "clcii"

import urllib,urllib2,re,sys,base64,socket,time #,pickle,collections
from xbmcswift2 import Plugin,xbmc,xbmcgui,ListItem,actions
from datetime import date
#from itertools import izip
import bibledata


socket.setdefaulttimeout(30)
plugin = Plugin('plugin.audio.audiobible')
forceViewMode=plugin.get_setting("forceViewMode", unicode)
viewMode=plugin.get_setting("viewMode", unicode)
basegatewayurl="http://www.biblegateway.com/audio/"

@plugin.route('/')
def enMain():
	items = [
		{'label': 'My Audio Bible', 'path': plugin.url_for('listMyBible')},
		{'label': 'Wisdom and Songs of the Day', 'path': plugin.url_for('listWisdomAndSongs')},
		{'label': 'Audio Bible Browser', 'path': plugin.url_for('listRecordings')}
	]
	recording=plugin.get_setting("defaultrecording", unicode)
	if forceViewMode=="true":
		xbmc.executebuiltin('Container.SetViewMode('+viewMode+')')
	return plugin.finish(items)
@plugin.route('/listMyBible/')
def listMyBible():
	defaultrecording=plugin.get_setting("defaultrecording",unicode)
	plugin.redirect('plugin://plugin.audio.audiobible/Browser/'+defaultrecording+'/')
@plugin.route('/Browser/')
def listRecordings():
	recordings=[]
	defaultrecording=str(plugin.get_setting("defaultrecording", unicode))
	alllanguages = sorted(bibledata.languagelib)
	for language in alllanguages:
		thislanguage = bibledata.languagelib[language]
		recordings.append({'label': thislanguage['title'], 
                            'path': plugin.url_for('listRecordings'),
                            'icon': 'DefaultShortcut.png'})
		for index in bibledata.recordinglib:
			recording = dict(bibledata.recordinglib[index])
			if recording['language']==language:
				alabel = recording['title']
				if recording['code']==defaultrecording:
					alabel=alabel + " (Current Default Recording)"
				recordings.append(
				{'label': alabel, 
				'path': plugin.url_for('listBooks', recording=recording['code']),
				'context_menu':[
					makedefaultrecording(recording['code'])
				],
				'icon': 'DefaultArtist.png'
				}
				)
		if forceViewMode=="true":
			xbmc.executebuiltin('Container.SetViewMode('+viewMode+')')
	return plugin.finish(recordings)
@plugin.route('/Browser/<recording>/')
def listBooks(recording):
	thisrecording = bibledata.recordinglib[recording]
	print(recording)
	books=[]
	allBooks=bibledata.booklib
	codes = dict(bibledata.bookorderlib)
	#for code, title in match:
	for item in codes:
 		code=codes[item]
 		book=dict(allBooks[code])
 		title=book['title']
		#or (bibledata.pluswisdom[recording] <> '' and (code=='Prov' or code=='Ps'))
 		if book['section']==thisrecording['section'] or thisrecording['section']=='ALL' or (thisrecording['pluswisdom'] == 1 and (code=='Prov' or code=='Ps')):
 			books.append(
 			{'label': title, 
 			'path': plugin.url_for('listChapters',recording=recording, book=code), 
 		 	'icon': 'DefaultPlaylist.png'}
 			)
	if forceViewMode=="true":
		xbmc.executebuiltin('Container.SetViewMode('+viewMode+')')
	return plugin.finish(books)
@plugin.route('/Browser/<recording>/<book>')
def listChapters(recording, book):
	thisrecording = bibledata.recordinglib[recording]
	thisbook=bibledata.booklib[book]
	chapters=thisbook['chapters']
	booktitle=thisbook['title']
	chapterslist=[]
	for x in range(1, chapters+1):
			title=booktitle +" - Chapter "+ str(x)
			chaptercode=book +"."+str(x)
			chapterslist.append(
				{'label': title, 
				'path': plugin.url_for('playChapter', 
				recording=recording, 
				book=book, 
				chapter=chaptercode),
				'icon': 'DefaultAudio.png',
				'info': ('audio'),
				'is_playable': True
				}
				)
	return plugin.finish(chapterslist)
@plugin.route('/listWisdomAndSongs/')
def listWisdomAndSongs():
	recording=plugin.get_setting("defaultrecording", unicode)
	thisrecording = bibledata.recordinglib[recording]
	chapterslist=[]	
	if thisrecording['section']=='NT' and thisrecording['pluswisdom']==0:
		chapterslist.append(
		{'label': 'Psalms and Proverbs not available with this default recording.', 
		'path': plugin.url_for('listMyBible'), 
		'icon': 'DefaultAudio.png',
		'info': ('audio'),
		'is_playable': False
		})
	else:
		today=date.today()
		thisday=today.day

		#Add Proverbs
		book='Prov'
		title='Proverbs '+ str(thisday)
		chaptercode=book+'.'+str(thisday)
		chapterslist.append(
				{'label': title, 
				'path': plugin.url_for('playChapter', 
				recording=recording, 
				book=book, 
				chapter=chaptercode),
				'icon': 'DefaultAudio.png',
				'info': ('audio'),
				'is_playable': True
				}
				)
	#Add Psalms
		if thisday==31:
			book='Ps'
			psalm=str(119)
			title='Psalms '+ psalm
			chaptercode=book+'.'+psalm
			chapterslist.append(
			{'label': title, 
			'path': plugin.url_for('playChapter', 
			recording=recording, 
			book=book, 
			chapter=chaptercode),
			'icon': 'DefaultAudio.png',
			'info': ('audio'),
			'is_playable': True
			}
				)
		else:
			for x in range(1, 6):
				book='Ps'
				psalm=str((thisday*5) - (5-x))
				title='Psalms '+ psalm
				chaptercode=book+'.'+psalm
				chapterslist.append(
								{'label': title, 
								'path': plugin.url_for('playChapter', 
													recording=recording, 
													book=book, 
													chapter=chaptercode),
								'icon': 'DefaultAudio.png',
								'info': ('audio'),
								'is_playable': True
								}
								)
	return plugin.finish(chapterslist)
@plugin.route('/Play/<recording>/<book>/<chapter>')
def playChapter(recording, book, chapter):
	try:
		chapter_url=basegatewayurl+getrecordingpath(recording)+chapter
		mp3path=getMp3link(chapter_url)
		bookinfo=bibledata.booklib[book]
		title=bookinfo['title'] + ' ' + chapter[:chapter.find('.')]
		extinfo={'title':title,'album':recording}
		listitem={'label': title,
				'path':mp3path,
				'icon':'DefaultAudio.png',
				'info_type':'audio',
				'info':extinfo,
				'is_playable': True
		}  
		return plugin.set_resolved_url(listitem)
	except ValueError:
		dialog = xbmcgui.Dialog()
		dialog.ok("Unable to play recording")
		xbmc.PlayerControl(Stop)
		
@plugin.route('/setdefaultrecording/<code>')
def setdefaultrecording(code):
	plugin.set_setting('defaultrecording',code)
	plugin.log.debug('current default'+code)
	plugin.redirect('plugin://plugin.audio.audiobible/')
	xbmc.executebuiltin('Container.Refresh()')
#functions	
def makedefaultrecording(code):
	label = 'Set Default Recording'
	new_url = plugin.url_for('setdefaultrecording', code=code)
	return(label, actions.background(new_url))

def parameters_string_to_dict(parameters):
        ''' Convert parameters encoded in a URL to a dict. '''
	paramDict = {}
	if parameters:
		paramPairs = parameters[1:].split("&")
		for paramsPair in paramPairs:
			paramSplits = paramsPair.split('=')
			if (len(paramSplits)) == 2:
				paramDict[paramSplits[0]] = paramSplits[1]
	return paramDict

def getUrl(url):
	try:
		req = urllib2.Request(url)
		req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 6.1; rv:18.0) Gecko/20100101 Firefox/18.0')
		response = urllib2.urlopen(req)
		link=response.read()
		response.close()
		return link
	except ValueError:
		dialog.xbmcgui.Dialog()
		dialog.ok('Error','Unable to complete request')

def getMp3link(url):
	content = getUrl(url)
	content = content[content.find('<audio id="audio-player" autoplay="autoplay">'):]
	content = content[content.find('<source src="')+13:]
	content = content[content.find(''):]
	content = content[:content.find('">')]
	mp3path = content
	return mp3path
def getrecordingpath(url):
	reader=url[url.find("-")+1:]
	style=url[:url.find("-")]
	path = reader + "/" + style + "/"
	return(path)
if __name__=='__main__':
	try:
		import StorageServer
	except:
		import storageserverdummy as StorageServer
	cache = StorageServer.StorageServer('AudioBible',24)
	try:
		plugin.run()
	except IOError:
		plugin.notify('Network Error')